File: app/src/main/java/com/example/android_arch/aidl/RemoteService.java
package com.example.android_arch.aidl;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.os.RemoteException;
import android.util.Log;

import com.example.android_arch.IMsgInterface;

import utils.LogUtils;

public class RemoteService extends Service {
    public RemoteService() {
    }

    @Override
    public IBinder onBind(Intent intent) {
        return new Stub();
    }

    private class Stub extends IMsgInterface.Stub {
        @Override
        public void sendMsg() throws RemoteException {
            LogUtils.INSTANCE.d("sendMsg");
            Log.d("RemoteService", "sendMsg");
        }

        @Override
        public void getMsg() throws RemoteException {
            LogUtils.INSTANCE.d("getMsg");
            Log.d("RemoteService", "getMsg");
        }

        @Override
        public void addMsg() throws RemoteException {
            LogUtils.INSTANCE.d("addMsg");
            Log.d("RemoteService", "addMsg");
        }
    }
}

File: app/src/main/java/com/example/android_arch/aidl/RemoteService.kt
package com.example.android_arch.aidl

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.widget.Toast
import com.example.android_arch.aidl.IServiceConnectService.Stub

class RemoteService : Service() {
    var iService = object :Stub(){
        override fun connect() {
            Thread.sleep(5000)
        }

        override fun disConnect() {
            Toast.makeText(this@RemoteService,"disconnect",Toast.LENGTH_LONG).show()
        }

        override fun isConnected() {
            TODO("Not yet implemented")
        }
    }

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }
}